packages/vite/src/node/optimizer/index.ts

```diff
@@ -495,27 +509,70 @@ export function runOptimizeDeps(
   // the optimizedDepInfo.processing promise for each dep
 
   const qualifiedIds = Object.keys(depsInfo)
+  let cleaned = false
+  const cleanUp = () => {
+    if (!cleaned) {
+      cleaned = true
+      // No need to wait, we can clean up in the background because temp folders
+      // are unique per run
+      fsp.rm(processingCacheDir, { recursive: true, force: true }).catch(() => {
+        // Ignore errors
+      })
+    }
+  }
 
-  const createEmptyProcessingResult = () => ({
+  const succesfulResult: DepOptimizationResult = {
     metadata,
-    commit: async () => {},
-    cancel: async () => {},
-  })
+    cancel: cleanUp,
+    commit: async () => {
+      const dataPath = path.join(processingCacheDir, '_metadata.json')
+      fs.writeFileSync(
+        dataPath,
+        stringifyDepsOptimizerMetadata(metadata, depsCacheDir),
+      )
+
+      // Write metadata file, delete `deps` folder and rename the `processing` folder to `deps`
+      // Processing is done, we can now replace the depsCacheDir with processingCacheDir
+      // Rewire the file paths from the temporal processing dir to the final deps cache dir
+      const temporalPath = depsCacheDir + getTempSuffix()
+      const depsCacheDirPresent = fs.existsSync(depsCacheDir)
+
+      if (isWindows) {
+        depsCacheDirPresent && (await safeRename(depsCacheDir, temporalPath))
+        await safeRename(processingCacheDir, depsCacheDir)
+      } else {
+        depsCacheDirPresent && fs.renameSync(depsCacheDir, temporalPath)
+        fs.renameSync(processingCacheDir, depsCacheDir)
+      }
```

Comment:
tiny nit: would be nice to use `if (depsCacheDirPresent)` instead of `&&` for calling functions.

It would also be great to explain why we do a double rename too (maybe that can be explained above). IIUC the rename-rename-delete approach is faster than delete-rename to prevent potential concurrent conflict.